File: train.lua
-- Load dependencies
local optim = require 'optim';
local gnuplot = require 'gnuplot';
local image = require 'image';
local display = require 'display';
local hdf5 = require 'hdf5'
local cuda = pcall(require, 'cutorch'); -- Use CUDA if available
local hasCudnn, cudnn = pcall(require, 'cudnn'); -- Use cuDNN if available
util = paths.dofile('util/util.lua')
-- require 'dpnn';


--=============================================================================
-- Command-line options
--=============================================================================
local cmd = torch.CmdLine();
-- major parameters
cmd:option('-model', 'ConvAE', 'Model: AE|SparseAE|DeepAE|ConvAE|UpconvAE|DenoisingAE|Seq2SeqAE|VAE|CatVAE|WTA-AE');
cmd:option('-batchSize', 64, 'Batch size');
cmd:option('-epochs', 20, 'Training epochs');
-- data loading
cmd:option('-datasetPath', '', 'Path for dataset folder')
cmd:option('-nThreads', 2, '# of threads for data loading')
-- optimizer
cmd:option('-optimiser', 'adagrad | adam', 'Optimiser');
cmd:option('-learningRate', 0.01, 'Learning rate');
cmd:option('-weightDecay', 0.0005, 'Weight decay coefficient for regularization');
-- others
cmd:option('-denoising', 0, 'Use denoising criterion');
cmd:option('-mcmc', 0, 'MCMC samples');
cmd:option('-sampleStd', 1, 'Standard deviation of Gaussian distribution to sample from');
-- cpu / gpu
cmd:option('-cpu', 0, 'CPU only (useful if GPU memory is too low)');
-- control
cmd:option('-continue_train', 0, "if continue training, load the latest model: true, false")

local opt = cmd:parse(arg);
if opt.cpu then
	cuda = false;
end
if opt.model == 'DenoisingAE' then
	opt.denoising = false; -- Disable "extra" denoising
end
print(opt)


--=============================================================================
-- Set up Torch
--=============================================================================
print('Setting up');
torch.setdefaulttensortype('torch.FloatTensor');
torch.manualSeed(854);
if cuda then
	require 'cunn';
	cutorch.manualSeed(torch.random());
end

local epoch_tm = torch.Timer()
local tm = torch.Timer()
local data_tm = torch.Timer()

--=============================================================================
-- Load data
--=============================================================================
-- check file existance
local inputFileList = {}
for file in paths.files(opt.datasetPath, ".h5") do
	table.insert(inputFileList, file);
end
assert(nil ~= next(inputFileList), "There is no proper input file at " .. opt.datasetPath)

local XTrainFile = hdf5.open(paths.concat(opt.datasetPath, inputFileList[1]), 'r');
local dataDim = XTrainFile:read('/data'):dataspaceSize()
XTrainFile:close();

local sampleLength, sampleWidth, sampleHeight = dataDim[2], dataDim[3], dataDim[4]
local XTrain = torch.Tensor(opt.batchSize, sampleLength, sampleWidth, sampleHeight)

local function load_data_from_file(inputFileName)
	local readFile = hdf5.open(paths.concat(opt.datasetPath, inputFileName), 'r');
	local dim = readFile:read('/data'):dataspaceSize();
	local numSamples = dim[1];
	print(('Reading data from %s : %d samples'):format(
		inputFileName, numSamples))
	-- local data = readFile:read('/data'):all();
	local data = readFile:read('/data'):partial({1, 100}, {1, dim[2]}, {1, dim[3]}, {1, dim[4]});
	readFile:close();

	return data
end


--=============================================================================
-- Create model & loss function
--=============================================================================
local function weights_init(m)
	local name = torch.type(m)
	if name:find('Convolution') then
		m.weight:normal(0.0, 0.02)
		m.bias:fill(0)
	elseif name:find('BatchNormalization') then
		if m.weight then m.weight:normal(1.0, 0.02) end
		if m.bias then m.bias:fill(0) end
	end
end

local Model = require ('models/' .. opt.model);
Model:createAutoencoder(XTrain);

-- if opt.denoising then
-- 	Model.autoencoder:insert(nn.WhiteNoise(0, 0.5), 1); -- Add noise during training
-- end

local autoencoder = Model.autoencoder;
local encoder = Model.encoder;
print(encoder);

autoencoder:apply(weights_init);


local criterion = nn.MSECriterion()
local softmax = nn.SoftMax() -- Softmax for CatVAE KL divergence


--=============================================================================
-- Data buffer and GPU
--=============================================================================
local batchInput  = torch.Tensor(opt.batchSize, sampleLength, sampleHeight, sampleWidth)
local batchOutput = torch.Tensor(opt.batchSize, sampleLength, sampleHeight, sampleWidth)

if cuda then
	print('transferring to gpu...')
	require 'cunn'
	cutorch.setDevice(opt.gpu)
	
	-- data buffer
	batchInput  = batchInput:cuda()
	batchOutput = batchOutput:cuda()

	-- network
	if hasCudnn then
		-- Use cuDNN if available
		autoencoder = util.cudnn(autoencoder)
		encoder = util.cudnn(encoder)
	end
	autoencoder:cuda();

	-- loss function
	criterion:cuda();
	
	print('done')
else
	print('running model on CPU')
end

-- Get parameters
local params, gradParams = autoencoder:getParameters();

--=============================================================================
-- Create optimiser function evaluation
--=============================================================================
local feval = function(params)

	-- Zero gradients
	gradParams:zero()

	-- Reconstruction phase
	-- Forward propagation
	batchOutput = autoencoder:forward(batchInput); -- Reconstruction
	local loss = criterion:forward(batchOutput, batchInput); -- xHat = batchOutput
	-- Backpropagation
	local gradLoss = criterion:backward(batchOutput, batchInput);
	autoencoder:backward(batchInput, gradLoss);

	return loss, gradParams
end


--=============================================================================
-- Training
--=============================================================================
print('Training')
autoencoder:training()

optimState = {
   learningRate = opt.learningRate,
   weightDecay = opt.weightDecay,
}

local __, loss
local losses = {}
-- to make a consistant data size
local leftDataLength = 0


for epoch = 1, opt.epochs do
	epoch_tm:reset()

	local fileIndices = torch.randperm(#inputFileList);
	for k = 1, #inputFileList do
		local fIdx = fileIndices[k];

		data_tm:reset()
		local data = load_data_from_file(inputFileList[fIdx]);
		print(('Done: %3f secs'):format(data_tm:time().real))

		-- Permute data
		data = data:index(1, torch.randperm(data:size(1)):long())
		print(('total samples: %d'):format(data:size(1)))
		print(('left sample before loop: %d'):format(leftDataLength))

		for start = leftDataLength+1, data:size(1), opt.batchSize do
			print(('start: %d'):format(start))
			-- Get minibatch
			local loadSize = math.min(data:size(1) - start + 1, opt.batchSize - leftDataLength)
			print(('loaded samples: %d'):format(loadSize))

			local readySize = leftDataLength + loadSize;
			local loadedData = data:sub(start, start + loadSize -1);

			if readySize < opt.batchSize then
				if  0 == leftDataLength then
					-- save and skip
					batchInput:sub(1, loadSize):copy(loadedData);
				else
					batchInput:sub(leftDataLength+1, readySize):copy(loadedData);					
				end
				leftDataLength = readySize;
				print(('left sample: %d'):format(leftDataLength))
			else
				if leftDataLength > 0 then
					print(('left samples: %d, loaded sampled: %d'):format(leftDataLength, loadSize))
				end

				batchInput:sub(leftDataLength+1, readySize):copy(loadedData);
				leftDataLength = 0;
				
				-- Optimize
				print('optimize start')
				-- __, loss = optim.adagrad(feval, params, optimState)
				loss = {1, 1}
				print(('optimize end, current loss: %.7f'):format(loss[1]))

				losses[#losses + 1] = loss[1]
			end
		end
		print("========================================")
	end
	
	-- Plot training curve(s)
	local plots = {{'Autoencoder', torch.linspace(1, #losses, #losses), torch.Tensor(losses), '-'}}
	gnuplot.pngfigure('Training.png')
	gnuplot.plot(table.unpack(plots))
	gnuplot.ylabel('Loss')
	gnuplot.xlabel('Batch #')
	gnuplot.plotflush()

	print(('End of epoch %d / %d \t Time Taken: %.3f secs'):format(
            epoch, opt.epochs, epoch_tm:time().real))
end


-- --=============================================================================
-- -- Plot reconstructions
-- --=============================================================================
-- image.save('Reconstructions.png', torch.cat(image.toDisplayTensor(x, 2, 10), image.toDisplayTensor(xHat, 2, 10), 1))

-- if opt.model == 'AE' or opt.model == 'SparseAE' or opt.model == 'WTA-AE' then
-- 	-- Plot filters
-- 	image.save('Weights.png', image.toDisplayTensor(Model.decoder:findModules('nn.Linear')[1].weight:view(x:size(3), x:size(2), Model.features):transpose(1, 3), 1, math.floor(math.sqrt(Model.features))))
-- end

-- if opt.model == 'VAE' then
-- 	if opt.denoising then
-- 		autoencoder:training() -- Retain corruption process
-- 	end

-- 	-- Plot interpolations
-- 	local height, width = XTest:size(2), XTest:size(3)
-- 	local interpolations = torch.Tensor(15 * height, 15 * width):typeAs(XTest)
-- 	local step = 0.05 -- Use small steps in dense region of 2D Gaussian; TODO: Move to spherical interpolation?

-- 	-- Sample 15 x 15 points
-- 	for i = 1, 15  do
-- 		for j = 1, 15 do
-- 			local sample = torch.Tensor({2 * i * step - 16 * step, 2 * j * step - 16 * step}):typeAs(XTest):view(1, 2) -- Minibatch of 1 for batch normalisation
-- 			interpolations[{{(i-1) * height + 1, i * height}, {(j-1) * width + 1, j * width}}] = Model.decoder:forward(sample)
-- 		end
-- 	end
-- 	image.save('Interpolations.png', interpolations)

-- 	-- Plot samples
-- 	local output = Model.decoder:forward(torch.Tensor(15 * 15, 2):normal(0, opt.sampleStd):typeAs(XTest)):clone()

-- 	-- Perform MCMC sampling
-- 	for m = 0, opt.mcmc do
-- 		-- Save samples
-- 		if m == 0 then
-- 			image.save('Samples.png', image.toDisplayTensor(Model.decoder.output, 0, 15))
-- 		else
-- 			image.save('Samples (MCMC step ' .. m .. ').png', image.toDisplayTensor(Model.decoder.output, 0, 15))
-- 		end

-- 		-- Forward again
-- 		autoencoder:forward(output)
-- 	end
-- elseif opt.model == 'CatVAE' then
-- 	if opt.denoising then
-- 		autoencoder:training() -- Retain corruption process
-- 	end

-- 	-- Plot "interpolations"
-- 	local height, width = XTest:size(2), XTest:size(3)
-- 	local interpolations = torch.Tensor(Model.N * height, Model.k * width):typeAs(XTest)

-- 	for n = 1, Model.N do
-- 		for k = 1, Model.k do
-- 			local sample = torch.zeros(Model.N, Model.k):typeAs(XTest)
-- 			sample[{{}, {1}}] = 1 -- Start with first dimension "set"
-- 			sample[n] = 0 -- Zero out distribution
-- 			sample[n][k] = 1 -- "Set" cluster
-- 			interpolations[{{(n-1) * height + 1, n * height}, {(k-1) * width + 1, k * width}}] = Model.decoder:forward(sample:view(1, Model.N * Model.k)) -- Minibatch of 1 for batch normalisation
-- 		end
-- 	end
-- 	image.save('Interpolations.png', interpolations)

-- 	-- Plot samples
-- 	local samples = torch.Tensor(15 * 15 * Model.N, Model.k):bernoulli(1 / Model.k):typeAs(XTest):view(15 * 15, Model.N * Model.k)
-- 	local output = Model.decoder:forward(samples):clone()

-- 	-- Perform MCMC sampling
-- 	for m = 0, opt.mcmc do
-- 		-- Save samples
-- 		if m == 0 then
-- 			image.save('Samples.png', image.toDisplayTensor(Model.decoder.output, 0, 15))
-- 		else
-- 			image.save('Samples (MCMC step ' .. m .. ').png', image.toDisplayTensor(Model.decoder.output, 0, 15))
-- 		end

-- 		-- Forward again
-- 		autoencoder:forward(output)
-- 	end
-- end





-- ()()
-- ('') HAANJU.YOO

File: train.py
import argparse
import os
import random
import time
import numpy as np
import torch
import torch.nn as nn
import torch.nn.parallel
import torch.nn.init
import torch.utils.data
import torch.backends.cudnn as cudnn
import torch.optim as optim
from torch.autograd import Variable
from models import AE, VAE
from data import VideoClipSets


def debug_print(arg):
    if not options.debug_print:
        return
    print(arg)


# =============================================================================
# OPTIONS
# =============================================================================
parser = argparse.ArgumentParser(description='Detecting abnormal behavior in videos')

# model related ---------------------------------------------------------------
parser.add_argument('--model', type=str, required=True, help='AE | AE_NEW | VAE | VAE_NEW')
parser.add_argument('--nc', type=int, default=10, help='number of input channel. default=10')
parser.add_argument('--nz', type=int, default=200, help='size of the latent z vector. default=100')
parser.add_argument('--nf', type=int, default=64, help='size of lowest image filters. default=64')
parser.add_argument('--l1_coef', type=float, default=0, help='coef of L1 regularization on the weights. default=0')
parser.add_argument('--l2_coef', type=float, default=0, help='coef of L2 regularization on the weights. default=0')
# training related ------------------------------------------------------------
parser.add_argument('--batch_size', type=int, default=64, help='input batch size. default=64')
parser.add_argument('--epochs', type=int, default=5000, help='number of epochs to train for. default=25')
parser.add_argument('--max_iter', type=int, default=150000, help='number of iterations to train for. default=150,000')
parser.add_argument('--partial_learning', type=float, default=1,
                    help='ratio of partial data for training. At least one sample from each file. default=1')
parser.add_argument('--continue_train', action='store_true', default=False,
                    help='load the latest model to continue the training, default=False')
# data related ----------------------------------------------------------------
parser.add_argument('--dataset', type=str, required=True,
                    help="all | avenue | ped1 | ped2 | enter | exit. 'all' means using entire data")
parser.add_argument('--data_root', type=str, required=True, help='path to base folder of entire dataset')
parser.add_argument('--image_size', type=int, default=227, help='input image size (width=height). default=227')
parser.add_argument('--workers', type=int, default=2, help='number of data loading workers')
# optimization related --------------------------------------------------------
parser.add_argument('--optimiser', type=str, default='adam', help='type of optimizer: adagrad | adam')
parser.add_argument('--learning_rate', type=float, default=0.0002, help='learning rate. default=0.0002')
parser.add_argument('--weight_decay', type=float, default=0.0005,
                    help='weight decay coefficient for regularization. default=0.0005')
parser.add_argument('--beta1', type=float, default=0.5, help='beta1 for Adam optimizer. default=0.5')
# display related -------------------------------------------------------------
parser.add_argument('--display', action='store_true', default=False,
                    help='visualize things with visdom or not. default=False')
parser.add_argument('--display_freq', type=int, default=5, help='display frequency w.r.t. iterations. default=5')
# GPU related -----------------------------------------------------------------
parser.add_argument('--num_gpu', type=int, default=1, help='number of GPUs to use. default=1')
# network saving related ------------------------------------------------------
parser.add_argument('--save_freq', type=int, default=10,
                    help='network saving frequency w.r.t. epoch number. default=500')
parser.add_argument('--save_path', type=str, default='./training_result',
                    help='path to trained network. default=./training_result')
parser.add_argument('--save_name', type=str, default='', help='name for network saving')
# ETC -------------------------------------------------------------------------
parser.add_argument('--random_seed', type=int, help='manual seed')
parser.add_argument('--debug_print', action='store_true', default=False, help='print debug information')
# -----------------------------------------------------------------------------

options = parser.parse_args()

# seed
if options.random_seed is None:
    options.random_seed = random.randint(1, 10000)

if options.model.find('VAE') != -1:
    options.variational = True

print(options)


# =============================================================================
# INITIALIZATION PROCESS
# =============================================================================

cuda_available = torch.cuda.is_available()

torch.manual_seed(options.random_seed)
if cuda_available:
    torch.cuda.manual_seed_all(options.random_seed)

cudnn.benchmark = True

# network saving
save_path = os.path.join(options.save_path, options.model)
try:
    os.makedirs(save_path)
except OSError:
    debug_print('WARNING: Cannot make saving folder')
    pass

# visualization
win_loss = None
if options.display:
    from visdom import Visdom
    viz = Visdom()

# =============================================================================
# DATA PREPARATION
# =============================================================================

# set data loader
options.dataset.replace(' ', '')  # remove white space
dataset_paths = []
mean_images = {}
if 'all' == options.dataset:
    options.dataset = 'avenue|ped1|ped2|enter|exit'
if 'avenue' in options.dataset:
    dataset_paths.append(os.path.join(options.data_root, 'avenue', 'train'))
    mean_images['avenue'] = np.load(os.path.join(options.data_root, 'avenue', 'mean_image.npy'))
# TODO: tokenize dataset string with '|'

dataset = VideoClipSets(dataset_paths)
dataloader = torch.utils.data.DataLoader(dataset=dataset, batch_size=options.batch_size, shuffle=True,
                                         num_workers=options.workers)
print('Data loader is ready')

# streaming buffer
tm_buffer_set = time.time()
input_batch = torch.FloatTensor(options.batch_size, options.nc, options.image_size, options.image_size)
recon_batch = torch.FloatTensor(options.batch_size, options.nc, options.image_size, options.image_size)
debug_print('Stream buffers are set: %.3f sec elapsed' % (time.time() - tm_buffer_set))

if cuda_available:
    debug_print('Start transferring to CUDA')
    tm_gpu_start = time.time()
    input_batch = input_batch.cuda()
    recon_batch = recon_batch.cuda()
    debug_print('Transfer to GPU: %.3f sec elapsed' % (time.time() - tm_gpu_start))

tm_to_variable = time.time()
input_batch = Variable(input_batch)
recon_batch = Variable(recon_batch)
debug_print('To Variable for Autograd: %.3f sec elapsed' % (time.time() - tm_to_variable))

viz_target_frame_index = int(options.nc / 2)


def pick_frame_from_batch(batch_data):
    return batch_data[0, viz_target_frame_index].cpu().numpy()


def gray_single_to_image(image):
    return np.uint8(image[np.newaxis, :, :].repeat(3, axis=0))


def sample_batch_to_image(batch_data):
    single_image = ((pick_frame_from_batch(batch_data) * 0.5) + 0.5) * 255  # [-1, +1] to [0, 255]
    # un-normalize
    return gray_single_to_image(single_image)


def decentering(image, dataset='avenue'):
    return gray_single_to_image(image * 255 + mean_images[dataset])


def draw_loss_function(win, losses, iter):
    cur_loss = np.zeros([1, len(losses.values())])
    x_values = np.ones([1, len(losses.values())]) * iter
    for i, value in enumerate(losses.values()):
        cur_loss[0][i] = value

    if win is None:
        legends = []
        for key in losses.keys():
            legends.append(key)
        win = viz.line(X=x_values, Y=cur_loss,
            opts=dict(
                title='losses at each iteration',
                xlabel='iterations',
                ylabel='loss',
                xtype='linear',
                ytype='linear',
                legend=legends,
                makers=False
            )
        )
    else:
        viz.line(X=x_values, Y=cur_loss, win=win, update='append')

    return win


def get_loss_string(losses):
    str_losses = 'Total: %.4f Recon: %.4f' % (losses['total'], losses['recon'])
    if 'variational' in losses:
        str_losses += ' Var: %.4f' % (losses['variational'])
    if 'l1_reg' in losses:
        str_losses += ' L1: %.4f' % (losses['l1_reg'])
    if 'l2_reg' in losses:
        str_losses += ' L2: %.4f' % (losses['l2_reg'])
    return str_losses

print('Data streaming is ready')


# =============================================================================
# MODEL & LOSS FUNCTION
# =============================================================================

# create model instance
# TODO: load pretrained model
if 'AE' == options.model:
    model = AE(num_in_channels=options.nc, z_size=options.nz, num_filters=options.nf)
elif 'VAE' == options.model:
    model = VAE(num_in_channels=options.nc, z_size=options.nz, num_filters=options.nf)
assert model
print(options.model + ' is generated')

# criterion
reconstruction_loss = nn.MSELoss()
variational_loss = nn.KLDivLoss()

# to gpu
if cuda_available:
    debug_print('Start transferring to CUDA')
    tm_gpu_start = time.time()
    model.cuda()
    reconstruction_loss.cuda()
    variational_loss.cuda()
    debug_print('Transfer to GPU: %.3f sec elapsed' % (time.time() - tm_gpu_start))

# for display
params = model.parameters()


def loss_function(recon_x, x, mu=None, logvar=None):
    # thanks to Autograd, you can train the net by just summing-up all losses and propagating them
    recon_loss = reconstruction_loss(recon_x, x)
    total_loss = recon_loss
    loss_info = {'recon': recon_loss.data[0]}

    if options.variational:
        assert mu is not None and logvar is not None
        # 0.5 * sum(1 + log(sigma^2) - mu^2 - sigma^2)
        kld_element = mu.pow(2).add_(logvar.exp()).mul_(-1).add_(1).add_(logvar)
        kld_loss = torch.sum(kld_element).mul_(-0.5)
        loss_info['variational'] = kld_loss.data[0]
        total_loss += kld_loss

    # TODO: check the effect of losses at below
    if 0.0 != options.l1_coef:
        l1_loss = options.l1_coef * torch.norm(params, 1)
        loss_info['l1_reg'] = l1_loss.data[0]
        # params.data -= options.learning_rate * params.grad.data
        total_loss += l1_loss

    if 0.0 != options.l2_coef:
        l2_loss = options.l2_coef * torch.norm(params, 2) ^ 2 / 2
        loss_info['l2_reg'] = l2_loss.data[0]
        total_loss += l2_loss

    loss_info['total'] = total_loss.data[0]

    return total_loss, loss_info


def save_model(filename):
    torch.save(model.state_dict(), '%s/%s' % (save_path, filename))
    print('Model is saved at ' + filename)


# =============================================================================
# TRAINING
# =============================================================================

print('Start training...')
model.train()

optimizer = optim.Adam(model.parameters(), lr=options.learning_rate, betas=(options.beta1, 0.999))

total_loss_history = []

tm_data_load_total = 0
tm_iter_total = 0
tm_loop_start = time.time()

# TODO: modify iter_count and epoch range with pretrained model's metadata
iter_count = 0
for epoch in range(options.epochs):
    tm_cur_iter_start = time.time()
    for i, data in enumerate(dataloader, 0):

        # data feed
        batch_size = data.size(0)
        input_batch.data.resize_(data.size()).copy_(data)
        recon_batch.data.resize_(data.size())

        # forward
        tm_train_start = time.time()
        model.zero_grad()
        recon_batch, mu_batch, logvar_batch = model(input_batch)
        # backward
        loss, loss_detail = loss_function(recon_x=recon_batch, x=input_batch, mu=mu_batch, logvar=logvar_batch)
        loss.backward()
        # update
        optimizer.step()
        tm_train_iter_consume = time.time() - tm_train_start

        # logging losses
        total_loss_history.append(loss.data[0])

        # visualize
        tm_visualize_start = time.time()
        if options.display:

            # visualize input / reconstruction pair
            viz_input_frame = decentering(pick_frame_from_batch(data))
            viz_input_data = sample_batch_to_image(data)
            viz_recon_data = sample_batch_to_image(recon_batch.data)
            viz_recon_frame = decentering(pick_frame_from_batch(recon_batch.data))
            if 0 == iter_count:
                print(viz_input_frame.shape)
                win_input_frame = viz.image(viz_input_frame, opts=dict(title='Input'))
                win_input_data = viz.image(viz_input_data, opts=dict(title='Input'))
                win_recon_data = viz.image(viz_recon_data, opts=dict(title='Reconstruction'))
                win_recon_frame = viz.image(viz_recon_frame, opts=dict(title='Reconstructed video frame'))
            else:
                viz.image(viz_input_frame, win=win_input_frame)
                viz.image(viz_input_data, win=win_input_data)
                viz.image(viz_recon_data, win=win_recon_data)
                viz.image(viz_recon_frame, win=win_recon_frame)

            # plot loss graphs
            win_loss = draw_loss_function(win_loss, loss_detail, iter_count)

        tm_visualize_consume = time.time() - tm_visualize_start

        iter_count += 1
        # # checkpoint w.r.t. iteration number
        # if 0 == iter_count % options.save_freq:
        #     save_model('%s_%s_iter_%03d.pth' % (options.dataset, options.model, iter_count))

        # TODO: save latest network with metadata containing saved network

        tm_iter_consume = time.time() - tm_cur_iter_start
        tm_cur_iter_start = time.time()  # to measure the time of enumeration of the loop controller, set timer at here

        # print iteration's summary
        print('[%02d/%02d][%04d/%04d] Iter:%06d %s'
              % (epoch+1, options.epochs, i, len(dataloader), iter_count, get_loss_string(loss_detail)))

        print('\tTime consume (secs) Total: %.3f CurIter: %.3f, Train: %.3f, Vis.: %.3f ETC: %.3f'
              % (time.time() - tm_loop_start, tm_iter_consume, tm_train_iter_consume, tm_visualize_consume,
                 tm_iter_consume - tm_train_iter_consume - tm_visualize_consume))

    print('====> Epoch %d is ternimated: Total loss is %f' % (epoch+1, total_loss_history[-1]))

    # checkpoint w.r.t. epoch
    if 0 == (epoch+1) % options.save_freq:
        save_model('%s_%s_epoch_%03d.pth' % (options.dataset, options.model, epoch))


#()()
#('')HAANJU.YOO

